app/src/main/cpp/skyline/kernel/scheduler.cpp

```diff
@@ -0,0 +1,367 @@
+                // We can yield the thread which is currently scheduled on the core by sending it a signal
+                // It is optimized to avoid waiting for the thread to yield on receiving the signal which serializes the entire pipeline
+                auto front{core.queue.front()};
+                front->forceYield = true;
+                core.queue.splice(std::upper_bound(core.queue.begin(), core.queue.end(), front->priority.load(), type::KThread::IsHigherPriority), core.queue, core.queue.begin());
+                core.queue.push_front(thread);
+
+                if (state.thread != front) {
+                    // If the inserting thread isn't at the front, we need to send it an OS signal to yield
+                    if (!front->pendingYield) {
+                        // We only want to yield the thread if it hasn't already been sent a signal to yield in the past
+                        // Not doing this can lead to races and deadlocks but is also slower as it prevents redundant signals
+                        front->SendSignal(YieldSignal);
+                        front->pendingYield = true;
+                    }
+                } else {
+                    // If the thread at the front is being yielded, we can just set the YieldPending flag
+                    // This avoids an OS signal and would cause a deadlock otherwise as the core lock would be relocked
+                    YieldPending = true;
+                }
+            } else {
+                core.queue.push_front(thread);
+            }
+            if (thread != state.thread)
+                thread->wakeCondition.notify_one(); // We only want to trigger the conditional variable if the current thread isn't inserting itself
+        } else {
+            core.queue.insert(nextThread, thread);
+        }
+    }
+
+    void Scheduler::MigrateToIdealCore(const std::shared_ptr<type::KThread> &thread, CoreContext *&core, std::unique_lock<std::mutex> &lock) {
+        // We need to check if the thread was in it's resident core's queue
+        // If it was, we need to remove it from the queue
+        auto it{std::find(core->queue.begin(), core->queue.end(), thread)};
+        bool wasInserted{it != core->queue.end()};
+        if (wasInserted) {
+            it = core->queue.erase(it);
+            if (it == core->queue.begin() && it != core->queue.end())
+                (*it)->wakeCondition.notify_one();
+        }
+        lock.unlock();
+
+        thread->coreId = thread->idealCore;
+        if (wasInserted)
+            // We need to add the thread to the ideal core queue, if it was previously it's resident core's queue
+            InsertThread(thread);
+
+        core = &cores.at(thread->coreId);
+        lock = std::unique_lock(core->mutex);
+    }
+
+    void Scheduler::WaitSchedule(bool loadBalance) {
+        auto &thread{state.thread};
+        CoreContext *core{&cores.at(thread->coreId)};
```

Comment:
No, there is nothing which can non-cooperatively change `coreId` for a thread directly. This can easily be implemented but it has additional performance implications of a conditional inside `wakeFunction` to check a variable for a new core ID.